File: src/contexts/graph/notes/domain/note.py
from typing import override

from src.contexts.graph.notes.domain.note_id import NoteId
from src.contexts.graph.notes.domain.note_title import NoteTitle
from src.contexts.graph.notes.domain.note_url import NoteUrl
from src.contexts.graph.notes.domain.related_notes_ids import RelatedNotesIds


class Note:
    _related_notes: RelatedNotesIds
    _title: NoteTitle
    _url: NoteUrl
    _id: NoteId

    def __init__(
        self,
        id_: NoteId,
        url: NoteUrl,
        title: NoteTitle,
        related_notes: RelatedNotesIds,
    ) -> None:
        self._id = id_
        self._url = url
        self._title = title
        self._related_notes = related_notes

    @classmethod
    def create(cls, id_: str, url: str, title: str, related_notes: list[str]) -> "Note":
        return Note(
            id_=NoteId(id_),
            url=NoteUrl(url),
            title=NoteTitle(title),
            related_notes=RelatedNotesIds(related_notes),
        )

    def to_primitives(self) -> dict:
        return {
            "id": self._id.value,
            "url": self._url.value,
            "title": self._title.value,
            "related_notes": self._related_notes.value,
        }

    @override
    def __repr__(self) -> str:
        return f"Note(id={self._id.value}, title={self._title.value})"

    @override
    def __eq__(self, other: "Note") -> bool:
        return self._id == other._id

File: backend/contexts/notes/src/main/kotlin/com/dimanupy/notes/note/domain/Note.kt
package com.dimanupy.notes.note.domain

data class NotePrimitives(
    val notionId: String,
    val title: String,
    val url: String,
    val relatedNotes: List<String>
)

data class Note private constructor(
    private val notionNoteId: NotionNoteId,
    private val title: NoteTitle,
    private val url: NoteUrl,
    private val relatedNotes: NoteRelatedNotes,
) {
    companion object {
        fun fromPrimitives(notionId: String, title: String, url: String, relatedNotes: List<String>): Note = Note(
            notionNoteId = NotionNoteId(notionId),
            title = NoteTitle(title),
            url = NoteUrl(url),
            relatedNotes = NoteRelatedNotes.create(relatedNotes),
        )
    }

    fun toPrimitives(): NotePrimitives = NotePrimitives(
        notionId = notionNoteId.value,
        title = title.value,
        url = url.value,
        relatedNotes = relatedNotes.value.map { it.value }
    )
}
